src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/DefaultRazorParsingPhase.cs

```diff
@@ -18,7 +25,27 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument, Cancellation
 
         foreach (var import in codeDocument.Imports)
         {
-            importSyntaxTrees.Add(RazorSyntaxTree.Parse(import, options));
+            if (!s_importTrees.TryGetValue(import, out var tree)
+                || !tree.Options.Equals(options))
+            {
+                tree = RazorSyntaxTree.Parse(import, options);
+
+#if NET
+                s_importTrees.AddOrUpdate(import, tree);
+#else
+                try
+                {
+                    // good effort update of CWT value
+                    s_importTrees.Remove(import);
+                    s_importTrees.Add(import, tree);
+                }
+                catch (ArgumentException)
+                {
+                }
+#endif
+            }
```

Comment:
In the case of default imports, it's possible for the same `RazorSourceDocument` to be returned across compilations (see `DefaultImportProjectItem`). For those scenarios, this represents a race condition if multiple razor files are compiled in parallel. Consider locking on `s_importTrees`.